INSTALL.rst

```diff
@@ -74,11 +75,13 @@ installed along with the system toolchain:
 
    apt-get install flex bison gcc python3 python3-pip
 
-The Python dependencies are installed using:
+All of the Python dependencies (build, run, and development) can be installed
+using:
 
 .. code:: sh
 
-   pip3 install --user -r requirements.txt
+   pip3 install --user pip-tools
+   pip3 install --user -r <(pip-compile --all-build-deps --all-extras --no-strip-extras 2>&1)
```

Comment:
This looks non-standard. The problem seems to be that we're not using a `requirements.txt` anymore. Probably because the dependencies are specified elsewhere.

Would it be worth automatically generating `requirements.txt` and add it to the Git repo? One then adds CI to check that the file is up to date. After doing so, installing the dependencies becomes rather standard:
```
pip install -r requirements.txt
```
(assuming a venv is active).

Reply:
All of the requirements are in `pyproject.toml` now, either as build-time, run-time, or optional (testing and docs for now, we can also add development-centric stuff like linters or type checkers in another category for instance).
Note that `pip-compile --all-build-deps --all-extras --no-strip-extras` actually generates a `requirements.txt`, but it's a very lenghty one (has all of the transitive dependencies I think).
On the other hand, `pip` has no way of actually _showing_ the dependencies/generating a `requirements.txt` from the TOML file, which is IMO a design issue on their end.
I've contemplated rolling our own script which just uses `pip`s internally vendored TOML reader to read the requirements, but it seems a bit convoluted for what should've been basic built-in functionality.

Reply:
Side note: It took 30s or more to run. It also creates a `requirements.txt` with precise version numbers.

Reply:
Yes, it pins everything and there isn't a way to remove this behavior. I don't really know of a good way to solve this except rolling our own parser, none of the standard tools seem to have a way of just parsing the TOML file and outputting the various dependencies in a `requirements.txt` format.